pkg/util/containers/v2/metrics/collector_ecs_fargate.go

```diff
@@ -0,0 +1,193 @@
+	metricsProvider.registerCollector(collectorMetadata{
+		id:       ecsFargateCollectorID,
+		priority: 0,
+		runtimes: allLinuxRuntimes,
+		factory:  func() (Collector, error) { return newEcsFargateCollector() },
+	})
+}
+
+type ecsFargateCollector struct {
+	client         *v2.Client
+	lastScrapeTime time.Time
+}
+
+// ecsStatsFunc allows mocking the ecs api for testing.
+type ecsStatsFunc func(ctx context.Context, id string) (*v2.ContainerStats, error)
+
+// newEcsFargateCollector returns a new *ecsFargateCollector.
+func newEcsFargateCollector() (*ecsFargateCollector, error) {
+	client, err := metadata.V2()
+	if err != nil {
+		return nil, err
+	}
+
+	return &ecsFargateCollector{client: client}, nil
+}
+
+// ID returns the collector ID.
+func (e *ecsFargateCollector) ID() string { return ecsFargateCollectorID }
+
+// GetContainerStats returns stats by container ID.
+func (e *ecsFargateCollector) GetContainerStats(containerID string, cacheValidity time.Duration) (*ContainerStats, error) {
+	stats, err := e.stats(containerID, cacheValidity, e.client.GetContainerStats)
+	if err != nil {
+		return nil, err
+	}
+
+	return convertEcsStats(stats), nil
+}
+
+// GetContainerNetworkStats returns network stats by container ID.
+func (e *ecsFargateCollector) GetContainerNetworkStats(containerID string, cacheValidity time.Duration, networks map[string]string) (*ContainerNetworkStats, error) {
+	stats, err := e.stats(containerID, cacheValidity, e.client.GetContainerStats)
+	if err != nil {
+		return nil, err
+	}
+
+	return convertEcsNetworkStats(stats.Networks, networks), nil
+}
+
+// stats returns stats by container ID, it uses an in-memory cache to reduce the number of api calls.
+// Cache expires every 2 minutes and can also be invalidated using the cacheValidity argument.
+func (e *ecsFargateCollector) stats(containerID string, cacheValidity time.Duration, clientFunc ecsStatsFunc) (*v2.ContainerStats, error) {
+	refreshRequired := e.lastScrapeTime.Add(cacheValidity).Before(time.Now())
+	cacheKey := fmt.Sprintf(statsCacheKey, containerID)
+	if cacheStats, found := cache.Cache.Get(cacheKey); found && !refreshRequired {
+		stats := cacheStats.(*v2.ContainerStats)
+		log.Debugf("Got ecs stats from cache for container %s", containerID)
+		return stats, nil
+	}
+
+	stats, err := clientFunc(context.TODO(), containerID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get container stats for %s: %w", containerID, err)
+	}
+
+	log.Debugf("Got ecs stats from ECS API for container %s", containerID)
+	e.lastScrapeTime = time.Now()
+	cache.Cache.Set(cacheKey, stats, statsCacheExpiration)
+
+	return stats, nil
+}
+
+func convertEcsStats(ecsStats *v2.ContainerStats) *ContainerStats {
+	return &ContainerStats{
+		Timestamp: time.Now(),
+		CPU:       convertCPUStats(&ecsStats.CPU),
+		Memory:    convertMemoryStats(&ecsStats.Memory),
+		IO:        convertIOStats(&ecsStats.IO),
+	}
+}
+
+func convertCPUStats(cpuStats *v2.CPUStats) *ContainerCPUStats {
```

Comment:
In theory, `cputStat` could be nil and same for any other field that's a pointer. I'm not sure whether it can happen or not in practice, but I have a similar situation in the provider that I'm implementing and I chose to check for nils just to be on the safe side. This applies to `cputStat.Usage` too and the parameters in similar functions below.

Reply:
See the [util function](https://github.com/DataDog/datadog-agent/blob/be59410d2c6d6321289287d8563e2c17493f369f/pkg/util/ecs/metadata/v2/client.go#L51), it cannot return nil stats

```
	if s, ok := stats[id]; ok && s != nil {
		return s, nil
	}
```

And the result type `ContainerStats` doesn't contain pointer fields either.

I'll add some checks anyways, just in case.